### ISSUE TITLE
Double-free-like issue with cleanups

### ISSUE THREAD

**Issue by eleses** [supercollider/supercollider#4806]
## Environment

* SuperCollider version: 3.10.4
* Operating system: doesn't matter
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce

```supercollider
(Pbind() <> Pfset({}, Pbind(\dur, Pseq([1], 1)), { "Cleanup!".postln })).play;
```

## Expected vs. actual behavior

Should print "Cleanup!" only once (as it happens if you don't Pchain something in front), but does it twice.

## Side comment
Pproto has the same problem internally, i.e. it executes its self-generated cleanups twice. It's not as easy to showcase that because Pproto has another bug #1732 that makes it completely ignore user-generated cleanup, but here's a way to do it by changing its default cleanup action (for groups in this case):

```
Event.addEventType(\customCleanupEvent, { "Custom (cleanup) event executed.".postln } );
EventTypesWithCleanup.cleanupTypes[\group] = \customCleanupEvent;
Pproto({ ~newgroup = (type: \group).yield;}, Pbind(\dur, Pseq([1], 1))).play;
```

I'm posting this a single bug because it might be fixample at EventStreamCleanup level with some kind of dupe-elimination logic. E.g. an instance of EventStreamCleanup could have a notion of parent so that it be aware of where in the stream chain it is positioned. I realize that EventStreamCleanup  is massively used in the Pattern library so any changes to that class could mean lots of regression issues.

**Comment by jamshark70** [comment]
There's an informal convention in the pattern library for event patterns to copy the input event before modifying it, but (first sign of trouble) this isn't done consistently: Pbind (`event = inevent.copy`) and Pchain (`inevent = inval.copy`) do, but Pfset doesn't. Also some patterns (accidentally saw `Plag`) copy at the beginning of `embedInStream` only, while others (Pbind, Pchain) copy within the loop.

I think this is intended to prevent accidentally modifying an event prototype. But that's unlikely to happen in common use because `Event.default` returns a unique copy, and EventStreamPlayer passes in a copy at the top level. So it's probably the case that most of these event-prototype copies are unnecessary.

The copy breaks EventStreamCleanup in the above case as follows:

- Pchain copies the input event: now there are `inval` (original) and `inevent` (copy).
- Pchain calls the sub-streams in reverse order, passing the copy.
- When Pfset runs out of events, its `cleanup.exit` does the cleanup function, adds the function into the event's `removeFromCleanup`, and returns the event to the calling routine.
- The Pfset routine has run out of things to do, so the active `next` call returns nil.
- Back in Pchain, `inevent` (copy) is set to nil (result of `str.next(inevent)`). So the reference to the event copy is lost.
- The only remaining reference to the incoming event is the original. But this wasn't passed down into Pfset, so it doesn't know about the `removeFromCleanup` value -- so it doesn't remove that cleanup function and the next level upstream will execute it again.

It can be fixed in two ways:

- Set a policy that `embedInStream` should avoid copying the input value unless absolutely necessary. I believe the impact of this would be minimal, but it would have to be tested thoroughly.
- Or, use a temporary variable for child stream results. If I change Pchain's `embedInStream` like this, then there is no double cleanup. This is safer (doesn't touch any broad assumptions), but it's harder to explain why it has to be written this way.

```supercollider
	embedInStream { arg inval;
		var streams, inevent, cleanup = EventStreamCleanup.new;
		var localEvent;
		streams = patterns.collect(_.asStream);
		loop {
			inevent = inval.copy;
			localEvent = inevent;
			streams.reverseDo { |str|
				localEvent = str.next(inevent);
				if(localEvent.isNil) { ^cleanup.exit(inevent) };
			};
			cleanup.update(localEvent);
			inval = yield(localEvent);
		};
	}
```

(Btw I've always wondered why the names of variables within patterns' `embedInStream` methods seem to be used almost randomly -- there are even cases of `event` meaning the input event and `inevent` meaning the result of the child stream!)

**Comment by jamshark70** [comment]
On second thought, I think we do have to forbid inevent copy globally.

The localized fix works here because Pfset happens to be one of those that doesn't copy the inevent. But if the sub-stream copies, and returns an event not identical to the one passed in, then I think you'll see the same problem.

While it's a bit risky and would call for testing, it's also the case that some patterns *don't* copy and we haven't seen ill effects from that. So it may not be so bad.

**Comment by eleses** [comment]
I tried quickly your Pchain code as a separate class and it seems you broke the chain...
```
p = Pchain (Pfuncn({|x| x + 1}, inf), Pseq([3, 6], 1))
p.asStream.next
```
Works with the original, but throws an error with yours...

**Comment by jamshark70** [comment]
Oh, jeez, that was dumb.

```
localEvent = str.next(inevent);
```

Should be

```
localEvent = str.next(localEvent);
```

But it doesn't matter, because you'll hit the same double cleanup bug if one of the chained patterns is one of those that does copy the inevent. So it's not a valid solution anyway (only illuminates the condition under which the problem occurs).

**Comment by eleses** [comment]
I think we're "back to square one" with that though because you're overwriting localEvent (which the sub-pattern hopefully modified with the cleanup-removal info) with the nil returned from the sub-pattern...

What that should look like is

```
var tmpEvent = str.next(localEvent);
cleanup.update(localEvent) // unconditional!!
if (tmpEvent.isNil) { ^cleanup.exit(inevent) }
```

I have yet to try this, but stay tuned...

**Comment by jamshark70** [comment]
> I think we're "back to square one" with that though because you're overwriting localEvent

No, my version keeps *two* references to the copy. `localEvent` becomes nil when the stream ends, but `inevent` still has a reference to the event that was passed into the child stream.

```
event = passed in from parent
\__ copy = inevent
    \__ 2nd reference, not a copy = localEvent
```

We pass localEvent down, and it gets modified... and inevent is another reference to the same object, so it knows everything that localEvent knows.

**Comment by eleses** [comment]
Yes, I stand corrected, your "v2" also works.

**Comment by jamshark70** [comment]
Cross-posted at https://scsynth.org/t/generalized-conceptual-design-flaw-double-free-of-cleanups/1843/7?u=jamshark70

So this post -- https://github.com/supercollider/supercollider/pull/4736#issuecomment-596803787 -- got me thinking overnight.

**Cleanup is about side effects**

Pure functional streams don't need cleanup because they don't make any messes.

With EventStreamPlayer, the whole point is the side effect of playing the events. And, you want to be able to stop the side effects on demand (allowing for differences of opinion whether "stop" means only to stop *new* side effects, or to stop the currently-active side effects immediately). Cleanup needs to work in both directions: bottom-up (the stream stopped, and needs to clean up its own stuff) *and* top-down (the user ordered the whole thing to stop).

The above-referenced github comment seems to be proposing that a pure data stream, e.g. `Pseries(0, 1, 10)`, should be just as "clean-uppable" as an event stream. Which raises two questions: 1/ What is the side effect to clean up? 2/ What is the concept of stopping a pure data stream?

1/ In theory, you could have a `Pfset({ set some state }, Pseries(...), { clear the state })`, so, OK. (It's not pure anymore, but OK.)

2/ We currently "stop" a pure data stream passively, by simply not requesting any more values. You *can* call `stop` on any routine at any time and it will yield `nil` after that point, but in practice, we don't bother.

#2 suggests that maybe it should be the stream itself handling the cleanup.

**Data streams with side-effects**

```
(
r = Routine {
	var synth;
	loop {
		synth = Synth(\default, [freq: exprand(200, 800)]);
		0.4.wait;
		synth.release;
		0.1.wait;
	}
}.play;
)

r.stop;
```

Common usage pattern here. We can think of this as a data stream (0.4, 0.1, 0.4, 0.1...) with side effects.

There's an 80% chance of hitting `stop` between `Synth()` and `release` -- leaving a hanging node. So that's a second reason to consider that maybe the proper home of the cleanup is the stream itself.

What if you could write something like this?

```
(
r = Routine {
	loop {
		var synth;
		var cleanup = { synth.release };
		synth = Synth(\default, [freq: exprand(200, 800)]);
		thisThread.addCleanup(cleanup);
		0.4.wait;
		thisThread.removeCleanup(cleanup);
		synth.release;
		0.1.wait;
	}
}.play;
)
```

And imagine something like that throughout the pattern library.

**Implementation**

I'm afraid I don't have time to build this myself at the moment, but some thoughts:

- Don't put this into Routine. There are other types of Streams (FuncStream, StreamCollect etc.). IMO a "clean-upabble Stream" should be a wrapper class.

- If you wrap a Routine in some other class, `thisThread` will be the Routine rather than the object handling cleanup. So you would need some other mechanism to get the current stream-with-cleanup. I don't know what that should be right now.

- For nested streams (call `next` on A, and A calls `next` on B, and B has a cleanup), I believe propagation upward would work something like:
  ```
  + StreamWithCleanup {
  	var stream;
  	next { |inval, cleanupDelta|
  		var childCleanupDelta = IdentityDictionary.new;
  		var result = stream.next(inval, childCleanupDelta);
  		... update my own cleanup according to delta...
  		... add child cleanups into passed-in cleanupDelta...
  		^result
  	}
  }
  ```
  ... and returning the result to the caller would give any parent streams the opportunity to update, and so on. I'm not sure if that's a good way to pass cleanup changes around, though.

Hope that helps. It may be difficult to implement but probably a design improvement. Comments (including "no, that's a terrible idea") welcome.

**Comment by eleses** [comment]
I actually found a case that breaks (still double cleans) with your v2 implementation but not with mine:

```
(Pchain(Pchain(Pfset({}, Pbind(\dur, Pseq([1, 0.4], 1)), {"Cleanup!".postln; }))).play;)
```

**Comment by jamshark70** [comment]
OK, sure, if it's fixable, that's great.

I admit that I'm lost about the rest. I used all my time for this issue today on my previous message.

**Comment by eleses** [comment]
Just a heads up that I've been working an automated testing framework for this (as was suggested to me on the user forum). I've got some good and bad news. Quite a few FilterPatterns play nicely with cleanups, but even the v3 bugfixed Pchain from above doesn't do well in this combo:

```
{ |x| PchainV3(x, x) } <> { |x| Pn(x, 3) } inits: 6, cleanups: 8
```

By way of explanation of what that means: to the right of that chain and not printed my error message because insofar it's always the same "cleanup injector" is a Pfset which counts into different variables how many inits and cleanups it gets as callbacks.

`{ |x| Pn(x, 3) }` by itself does `inits: 3, cleanups: 3` as one might expect.

**Comment by eleses** [comment]
I figured out why something like `Pchain(Pn(Pfset (...), 2), Pn(Pfset (...), 2))` doesn't work properly and cannot work properly even with the v3 implementation of Pchain. Basically what happens is that Pchain via its inEvent modification contaminates the first Pn so that its inner Pfset executes the cleanup of both Pfsets, even though it's not supposed to know about the 2nd Pfset's cleanup set. So after that happens, the first Pn restarts its Pfset, but then 2nd Pn is stuck just before yielding nil, which it does precisely on the fist `next` call from the new instance of the leftmost Pfset... and since had (legitimately) accumulated (just) its own cleanup, it will execute it now.

Basically the localEvent (to use the var name in your code) acts as "bus" between all the Pchain substream contaminating them with cleanup info they are *not* supposed to know about.

This might be fixable, but it's a real mess with all these side effects...

**Comment by jamshark70** [comment]
While it's a good thing to fix odd behavior in cleanups:

- I do think it will be a lot simpler if cleanups are associated with stream instances. What if the EventStreamCleanup stored a dictionary of `stream -> cleanupFuncs`? Then, when a specific sub-thread is cleaning up, it can be guaranteed to touch only its own funcs.
  I have doubts about fixing these bugs without changing the data model. But so far this idea has not gained any traction.
- A generalized constructor-destructor concept for patterns must consider the case of resources that need to persist through multiple play/stop cycles. Because EventStreamCleanup clears resources upon stop, by definition it can't handle persistent resources -- therefore it can never be a fully general destructor. Sure, objectively it's better to fix bugs -- but these bugs are being encountered in part by trying to make patterns do something which (IMO) patterns should not be responsible for. I'd like to see some of the attention to arbitrary combinations of Pchain and Pfset be directed toward an architecture that makes it less necessary to think in terms of these arbitrary combinations.

In short, then, +1 to fixing bugs where you find them, but -1 to the idea that pattern objects by themselves can handle every case.

**Comment by eleses** [comment]
Yeah, the fixup of the present design by doing "zero-inEvent-copy" everywhere in the Pattern library is quite hairy in terms of reasoning about what's going on. I'm also afraid that eliminating the existing inEvent copies (which some patterns do and some don't copy), while not breaking anything in the existing TestPattern suite (e.g. Pchain v4 sill passes it), could break people's more complex code that somehow relies on inEvent copies as- and where they happen today.

Having a single queue of cleanups per "stream head" would be conceptually simpler than the (per-some-nodes) replicated structure that is in place today. There is such a "final" copy anyhow that `EventStreamPlayer` actually holds, which it calls `.terminate` on when its own `.stop` method gets called. There would be no need to propagate addToCleanup and removeFromCleanup messages if there's no replication. One Pattern, namely, PmonoArtic manipulates these addToCleanup/removeFromCleanup messages "manually" in addition to using the EventStreamCleanup helper functions, so it's a bit of a leaky abstraction.

I'll have to think some more about the implications of this simpler design though. (I have to go out the door in 5 minutes right now. More thoughts on this later....)

**Comment by eleses** [comment]
> What if the EventStreamCleanup stored a dictionary of stream -> cleanupFuncs? Then, when a specific sub-thread is cleaning up, it can be guaranteed to touch only its own funcs.

I'm not seeing exactly how this would work because cleanups can and in fact should get executed by "parent patterns" sometimes (or in a data-flow perspective, "downstream"). That's the whole point of replicating the cleanup function-sets to parent-pattern streams. A cleanup function can be legitimately executed by either the stream that created it or any stream that's pulling data from it, iff the the latter decides to stop pulling data because of its own internal logic. 

What you are proposing would work if the stream which stops pulling would also propagate some other signal so that the sub-stream from which it's not pulling anymore would cleanup itself. But this kind of propagation is currently not done for anything in SC, not even for stream resets let alone for "just not pulling data from you anymore"...

---

On the other hand, I think I know how to "have our cake and eat it" though, namely keep the current semi-haphazard inEvent copy calls where they happen, while at the same time creating a "dedicated bus" for cleanup messages attached to the event. The trick would be to simply include those in an inner-event 

```
e = (foo: 1, subE: (bar: 2))
e.identityHash.asHexString
d = e.copy
d.identityHash.asHexString // differs from e's

e[\subE].identityHash.asHexString 
d[\subE].identityHash.asHexString // same as e[\subE]
```

Basically, instead of putting the addToCleanup and removeFromCleanup directly in the [in]Event, as it currently happens, we'd put them in a dedicated inner-event, the reference to which happens to survive Event.copy. Luckily for us, no pattern does `deepCopy` on an event!

This inner event, let's call if `cleanupBus` for now (I prefer a name that emphasizes this stuff shouldn't be copied) would be created by the `play` method, which luckily is not overridden much if at all in the class library, i.e. all patterns seem to use the `play` they inherit from the abstract `Pattern` class.

Also, I'm aware that there exist `parent` and `proto` pointers in each Event, that could be used for the same purpose as my `cleanupBus`, but it's possible something else "messes up" with `proto` and `parent` (e.g nil's them or entirely rewrites them) and their (quite generic) names make it a bit hard to grep where they get used/changed. Debug-wise, it's also easier to see if a dedicated cleanup field/channel gets "dropped" somewhere.

There's still the problem that if you "manually" create a stream from pattern with `asStream`, it won't work properly cleanup-wise if you don't "buss it" with whomever it's supposed to communicate cleanup messages with. I don't see an easy automation for this right now since `asStream` doesn't take a proto-Event as an argument, i.e. you have pass "the right one" to every `next()` etc. which is basically what `EventStreamPlayer` does.

**Comment by jamshark70** [comment]
In another discussion, it came up that an idempotent function would prevent the action from firing twice, even if the function object is called twice.

We actually have an idempotent function: Thunk.

So, the least disruptive solution may be for EventStreamCleanup to wrap cleanup functions in Thunk.

Should be tested first but if it works, then the whole problem goes away by touching a couple of lines of code (rather than rearchitecting and introducing the risk of all-new subtle bugs).

**Comment by eleses** [comment]
@jamshark70: what happens if you copy a Thunk before it gets called the first time? Do the (two) copies run once or twice?

**Comment by jamshark70** [comment]
> You need an "uncopyable" Thunk to fix this.

Where are we copying cleanup functions?

**Comment by jamshark70** [comment]
```
(Pbind() <> Pfset({}, Pbind(\dur, Pseq([1], 1)), { "Cleanup!".postln })).play;

(Pchain(Pchain(Pfset({}, Pbind(\dur, Pseq([1], 1)), {"Cleanup!".postln; }))).play;)

~pf1 = Pfset({}, Pbind(\dur, Pseq([1], 1)), { "Cleanup 1!".postln });
~pf2 = Pfset({}, Pbind(\degree, Pseq([10], 1)), { "Cleanup 2!".postln });

Pchain(Pchain(~pf1, ~pf2)).play;
```

All of these do single cleanups when EventStreamCleanup does `function = Thunk(function);` before adding.

**Comment by jamshark70** [comment]
Apologies, I think I was a little too curt, and I didn't explain my reasoning.

What I mean is:

To say that we *need* an un-copyable Thunk to fix multiple evaluation of cleanup functions requires one of two conditions to be true: A/ that, somewhere in the class library now, some use of EventStreamCleanup copies a function passed to it (or, some client of EventStreamCleanup is copying a function), or B/ there is a legitimate scenario that could be handled only by copying a cleanup function.

If the logical expression `A or B` is false, then a copyable Thunk can solve the cleanup problem.

A is, I'm quite sure, false, and easily verified to be false by checking everywhere that EventStreamCleanup is used.

For B: I can't think of any valid cases. That isn't *proof* necessarily, but so far, we've been able to handle all cleanup cases by passing open functions -- that is, the function literal exists within the scope where the cleanup is being performed, and there is nowhere that we do `cleanupFunction = { ... }` in one place and `cleanup.addFunction(cleanupFunction)` in another place (which is not even copying! only setting up a situation where one might be tempted to copy).

(Where would the function be copied, anyway? The pattern is stateless, so, not there. A stream made from the pattern can't be copied.)

So, if it doesn't make sense to copy an event stream cleanup function, then a use of Thunk for event stream cleanup doesn't need to protect against copies.

FWIW this is what I tested:

```diff
diff --git a/SCClassLibrary/Common/Streams/EventStreamCleanup.sc b/SCClassLibrary/Common/Streams/EventStreamCleanup.sc
index c0922228d..a41c79503 100644
--- a/SCClassLibrary/Common/Streams/EventStreamCleanup.sc
+++ b/SCClassLibrary/Common/Streams/EventStreamCleanup.sc
@@ -14,6 +14,7 @@ EventStreamCleanup {
 
        addFunction { |event, function |
                if(event.isKindOf(Dictionary)) {
+                       function = Thunk(function);
                        functions.add(function);
                        event[\addToCleanup] = event[\addToCleanup].add(function);
                };
@@ -21,6 +22,7 @@ EventStreamCleanup {
 
        addNodeCleanup { |event, function |
                if(event.isKindOf(Dictionary)) {
+                       function = Thunk(function);
                        functions.add(function);
                        event[\addToNodeCleanup] = event[\addToNodeCleanup].add(function);
                };
```

I believe this is valid. Functions are removed only in EventStreamCleanup:update -- `functions.removeAll(event[\removeFromCleanup])` and `event[\removeFromCleanup]` is only ever populated from the `functions` Set -- so it will only ever be removing the specific Thunk instances created when adding the function.

**Comment by eleses** [comment]
Could work, not sure. Alas my SC time is pretty limited nowadays. Reminder that I wrote a mini test-suite with some stress tests, which could be used to test this or other ideas, at least to some extent

https://gist.github.com/eleses/abed028654f5736678e610a0aef6fe22

**Comment by jamshark70** [comment]
> why is there a separate \addToNodeCleanup distinct from \addToCleanup?

Probably something that considered but was never finished, or which turned out to be unnecessary.

> Reminder2: you'll also have to fix PmonoArticStream

Good point -- I had forgotten about that.

There's a way to do without `\removeFromCleanup` altogether. If Thunk can report `didEvaluate { ^function.isNil }`, then EventStreamCleanup `update` and `exit` can garbage-collect cleanups that already executed by: `functions = functions.reject(_.didEvaluate);`.

I tested this with PmonoArtic:

```
(
p = PmonoArtic(\default, \dur, 0.1);
q = PmonoArticStream(p);

r = Routine { |inval| q.embedInStream(inval) };
)

e = r.next(Event.default); e.play;
```

... and checking `q.slotAt(\cleanup).functions`, there is only ever one `[ a Thunk ]`.

There's a side benefit: PmonoArtic's manual-stop behavior becomes consistent, where it wasn't before:

```
p = Pmono(\default, \dur, 4).play;

p.stop;  // immediate

p = PmonoArtic(\default, \dur, 4, \legato, 1.01).play;

p.stop;  // immediate

p = PmonoArtic(\default, \dur, 4).play;

p.stop;  // current version: plays full-duration note
```

If you stop a PmonoArtic, the note is terminated immediately if it was going to sustain into the next, but if it wasn't going to sustain all the way, it waits until the scheduled release time. That's a bit dubious. With the change I just tested (taking out `\removeFromCleanup`), `p.stop` on Pmono *always* stops the note immediately. That's a slight breaking change, but I think beneficial (it's very odd that PmonoArtic isn't consistent with itself).

```diff
diff --git a/SCClassLibrary/Common/Core/Function.sc b/SCClassLibrary/Common/Core/Function.sc
index eb74af7c8..a6eb74002 100644
--- a/SCClassLibrary/Common/Core/Function.sc
+++ b/SCClassLibrary/Common/Core/Function.sc
@@ -337,18 +337,25 @@ Thunk : AbstractFunction {
        // a thunk is an unevaluated value.
        // it gets evaluated once and then always returns that value.
        // also known as a "promise" in Scheme.
-       // thunks have no arguments.
        var function, value;
 
        *new { arg function;
                ^super.newCopyArgs(function)
        }
-       value {
-               ^value ?? { value = function.value; function = nil; value }
+       value { |... args| ^this.prEvaluate(\valueArray, args) }
+       valueArray { |... args| ^this.prEvaluate(\valueArray, *args) }
+       valueEnvir { |... args| ^this.prEvaluate(\valueArrayEnvir, args) }
+       valueArrayEnvir { |... args| ^this.prEvaluate(\valueArrayEnvir, *args) }
+
+       didEvaluate { ^function.isNil }
+
+       prEvaluate { |selector, args|
+               ^value ?? {
+                       value = function.performList(selector, args);
+                       function = nil;
+                       value
+               }
        }
-       valueArray { ^this.value }
-       valueEnvir { ^this.value }
-       valueArrayEnvir { ^this.value }
 }
 
 UGenThunk : Thunk {
diff --git a/SCClassLibrary/Common/Streams/EventStreamCleanup.sc b/SCClassLibrary/Common/Streams/EventStreamCleanup.sc
index c0922228d..b31739b6a 100644
--- a/SCClassLibrary/Common/Streams/EventStreamCleanup.sc
+++ b/SCClassLibrary/Common/Streams/EventStreamCleanup.sc
@@ -14,23 +14,27 @@ EventStreamCleanup {
 
        addFunction { |event, function |
                if(event.isKindOf(Dictionary)) {
+                       function = Thunk(function);
                        functions.add(function);
                        event[\addToCleanup] = event[\addToCleanup].add(function);
                };
+               ^function  // give wrapped function back to user (see PmonoStream:prInitNode)
        }
 
        addNodeCleanup { |event, function |
                if(event.isKindOf(Dictionary)) {
+                       function = Thunk(function);
                        functions.add(function);
                        event[\addToNodeCleanup] = event[\addToNodeCleanup].add(function);
                };
+               ^function
        }
 
        update { | event |
                if(event.isKindOf(Dictionary)) {
                        functions.addAll(event[\addToNodeCleanup]);
                        functions.addAll(event[\addToCleanup]);
-                       functions.removeAll(event[\removeFromCleanup]);
+                       functions = functions.reject(_.didEvaluate);
                };
                ^event
        }
@@ -40,7 +44,7 @@ EventStreamCleanup {
                        this.update(event);
                        if(functions.notEmpty) {
                                functions.do(_.value(freeNodes) );
-                               event[\removeFromCleanup] = event[\removeFromCleanup].addAll(functions);
+                               functions = functions.reject(_.didEvaluate);
                        };
                        this.clear;
                };
diff --git a/SCClassLibrary/Common/Streams/PmonoStreams.sc b/SCClassLibrary/Common/Streams/PmonoStreams.sc
index 77ecba263..5b51f6114 100644
--- a/SCClassLibrary/Common/Streams/PmonoStreams.sc
+++ b/SCClassLibrary/Common/Streams/PmonoStreams.sc
@@ -68,7 +68,7 @@ PmonoStream : Stream {
                                if(event.isRest.not) {
                                        ~type = \monoNote;
                                        ~instrument = pattern.synthName;
-                                       cleanup.addFunction(event, currentCleanupFunc = { | flag |
+                                       currentCleanupFunc = cleanup.addFunction(event, { | flag |
                                                if (flag) { (id: id, server: server, type: \off,
                                                        hasGate: hasGate,
                                                        schedBundleArray: schedBundleArray,
@@ -129,7 +129,6 @@ PmonoArticStream : PmonoStream {
                                };
                                sustain = event.use { ~sustain.value };
                                if(sustain.notNil and: { sustain < event.delta }) {
-                                       event[\removeFromCleanup] = event[\removeFromCleanup].add(currentCleanupFunc);
                                        thisThread.clock.sched(sustain, {
                                                currentCleanupFunc.value(true);
                                                rearticulating = true;
```

**Comment by eleses** [comment]
> Pmono always stops the note immediately. 

Pmono always did that, as far as I recall. Did you mean PmonoArtic? The latter had/has this "split personality" that it behaves (on stop) like Pmono for legato >= 1 and like Pbind for legato < 1. (See https://scsynth.org/t/what-explains-pmono-vs-pbind-stop-behavior/1802)

> If you stop a PmonoArtic, the note is terminated immediately if it was going to sustain into the next, but if it wasn't going to sustain all the way, it waits until the scheduled release time. That's a bit dubious. 

Yes it is :-) I would prefer if it at least had a flag that turned off the Pbind-like behavior on legato < 1 because that's pretty easy to do. (Of course, an ineffective \removeFromCleanup line achieves the same, except is not configurable)

If you return the Thunked function from addFunction (as you do now), then the 
`removeFromCleanup` "lookup" works by matching the same Thunk. 

I haven't though carefully about your `didEvaluate` bit and the removal of `removeFromCleanup`, so no comments on that yet...

**Comment by jamshark70** [comment]
> Pmono always did that, as far as I recall. Did you mean PmonoArtic? The latter had/has this "split personality" that it behaves (on stop) like Pmono for legato >= 1 and like Pbind for legato < 1. (See https://scsynth.org/t/what-explains-pmono-vs-pbind-stop-behavior/1802).

Sorry that I was slightly unclear about that. I was intending to draw a contrast between Pmono and PmonoArtic behavior.

Pmono can only behave as it does now because there is no scheduled release time.

> Clearly though that was intentional because it takes that \removeFromCleanup line to make it work like that.

There was no other way to remove the cleanup function. The behavior is a side effect of that, and was simply not noticed previously.

I don't like the behavior.

**Comment by eleses** [comment]
> PmonoArtic  ... I don't like the behavior.

Ok with me, I wasn't exactly fond of the old behavior. I was just trying to keep the old semantics for the sake of not having one more point of possible object to a fix what is going to involve at least two semantic/breaking changes (Thunk.value args and EventStreamCleanup.addFunction return value). I think those are unlikely to objected to though.

As far as removing cleanups in general (not from PmonoArtic) I would suggest though keeping a way to "manually" remove cleanups too. In `EventStreamCleanup.update` I'm not seeing an issue keeping the old line that uses \removeFromCleanup and *adding* the new auto-removal for already executed Thunks, instead of replacing the old method. Since the cleanups are stored in a Set, a "double removal" will not cause any problems (although in the usual use case for \removeFromCleanup, I expect cleanups will be removed manually only before they are actually executed.)

```
z = Set.new
z.addAll([1, 2, 3])
z.removeAll([1, 2])
z.removeAll([1, 2]) // no err
z.addAll([1, 2, 3])
z.reject((_ < 3))
z.reject((_ < 3)) // no err
```

One could keep the \removeFromCleanup in `EventStreamCleanup.exit` too as harmless, but there it's pretty pointless as is happens after the cleanup is executed already... 

Alternatively, this was something I was considering with the "abandon" method in my sketch, we could have way to set the `didEvaluate` in Thunk *without* actually executing the stored function, as this will basically "abandon"/cancel its execution even if `value` is called later on.

**Comment by jamshark70** [comment]
> I would suggest though keeping a way to "manually" remove cleanups too. In EventStreamCleanup.update I'm not seeing an issue keeping the old line that uses \removeFromCleanup and adding the new auto-removal for already executed Thunks...

I wouldn't object to that -- fine to leave the option available.

**Comment by eleses** [comment]
Ok, so at least the two of us agree on most of this. :+1:  The practical question now is who starts making commits or pull requests.  (I guess we'll see more comments/input from others once that that happens.)  Also, should this be "chunked" as two separate PRs/commits on the main, one for )(the prerequisite) Thunk improvements and one for the EventStreamCleanup after that?

**Comment by eleses** [comment]
I've been considering a first PR on the (prerequisite) Thunk changes. But I'm hesitating now because I'm not sure a general-purpose Thunk is unobjectionable with these semantics, namely the fact that the code/function is frozen once the Thunk is created (as before), but the parameters become so only when the Thunk is called/evaluated for the first time. What if it's called with different parameters a 2nd time? Should that print some kind of warning?

It might be less objectionable if we created (say) CleanupThunk that works like we need it, since the only actual parameter is that `freeNodes` flag thing... and that's largely hidden from the user. The documentation only says/promises

> freeNodes	Used internally.

**Comment by telephon** [comment]
> I've been considering a first PR on the (prerequisite) Thunk changes. But I'm hesitating now because I'm not sure a general-purpose Thunk is unobjectionable with these semantics, namely the fact that the code/function is frozen once the Thunk is created (as before), but the parameters become so only when the Thunk is called/evaluated for the first time. What if it's called with different parameters a 2nd time? Should that print some kind of warning?

We've had a quick exchange on the mailing list already, I don't think you should change thunk, because passing another set of arguments when you call it again will produce confusing results. Calling it should be idempotent. But why not make a separate class? 

```
// naming follows EmbedOnce
CallOnce : Thunk {
	value { arg ... args;
		^value ?? { value = function.valueArray(args); function = nil; value }
	}
	valueArray { args ... args; ^this.valueArray(args) }
	valueEnvir { args ... args;  ^this.valueArray(args) }
	valueArrayEnvir { args ... args;  ^this.valueArray(args) }
}
```

**Comment by eleses** [comment]
@telephon Thank you. What I had already done (as James and I were discussing this thunk/do-once-based to cleanups) was to create a CleanupThunk class and used that in my EvenStreamCleanup fix, simply because I had not audited the classlib to see if it was never accidentally calling Thunk with parameters. (It turned out later when I did search for all occurrences that it doesn't.) So the solution you've suggested is already tested too :+1:  I should probably subscribe to the mailing list... Was it he users or is there a dev one?

**Comment by telephon** [comment]
ok, I'll do this.

**Comment by telephon** [comment]
Just for the note, we may have a similar issue with [OSCBundle](https://github.com/supercollider/supercollider/blob/develop/SCClassLibrary/Common/Control/OSCBundle.sc), there we also want to be able to cancel a scheduled message, and possibly avoid double calls.

But this is not in the scope of the current PR of course.

**Comment by jamshark70** [comment]
> Just for the note, we may have a similar issue with OSCBundle, there we also want to be able to cancel a scheduled message...

The issue with EventStreamCleanup is that cleanup functions are propagated through multiple instances, and because of some specific patterns' implemention (e.g. Pchain copies events in a way that Pbind doesn't), cleanup functions might not be reliably removed at parent/sibling levels.

If OSCBundle isn't used in a similar, multi-leveled way, it probably won't have the same problem.

**Comment by telephon** [comment]
> If OSCBundle isn't used in a similar, multi-leveled way, it probably won't have the same problem.

Yes, it isn't.
